66 new directory structure

dabele opened the pull request:
## Changes

Lots of files!

- New Directory structure:
```yaml
- cpp
    - memilio # used to be epidemiology
        - util # generic programming utility
        - math # math utility (integrators, linalg,...)
        - io # used to be in epidemiology_io, could also be in util but I think it deserves its own
        - epidemiology # non-model-specific utility for epidemiological models (abm or compartment)
        - compartments # compartment model base classes
        - mobility # non-model-specific spatial resolution, used to be migration
    - models
        - secir
        - abm
        - seir
    - examples 
    - simulations
    - tests
```
- removed epidemiology_io, HDF5 and jsoncpp are optional dependencies of memilio
- CI runs builds with and without optional dependencies (#75)
- improved modularity of main CMakeLists.txt (e.g. memilio subdirectory has its own CMakeLists)
- models are each their own library (but built in the same CMake project)
- added READMEs in some places where they were missing, moved some redundant info, main instruction for the c++ part is cpp/README.md
- renamed namespace `epi` to `mio`
- for unknown reasons there were a few new compiler warnings that had to be resolved, mostly with Clang; some warnings were disabled locally, some globally
- removed some dead code (e.g. secir.cpp)
- renamed runUnitTest to memilio-test (that was bothering me for a while, but not enough to change it :smile:)
- CMake install config probably doesn't work yet (maybe it never did, not sure anyone ever tested), created issue #89 

closes #66 
closes #75 

## Merge Request - GuideLine Checklist 

**Guideline** to check code before resolve WIP and approval, respectively.
As many checkboxes as possible should be ticked.

### Checks by code author:
* [x] There is at least one issue associated with the pull request.
* [x] The branch follows the naming conventions as defined in the [git workflow](git-workflow).
* [x] New code adheres with the [coding guidelines](coding-guidelines)
* [x] Tests for new functionality has been added
* [x] A local test was succesful
* [x] There is appropriate **documentation** of your work. (use doxygen style comments)
* [x] If new third party software is used, did you pay attention to its license? Please remember to add it to the wiki after successful merging.
* [ ] If new mathematical methods or epidemiological terms are used, has the glossary been updated ? Did you provide further documentation ?
 is present or referenced. Please provide your references.
* [ ] The following questions are addressed in the documentation*:  Developers (what did you do?, how can it be maintained?), For users (how to use your work?), For admins (how to install and configure your work?)
* [x] For documentation: Please write or update the Readme in the current working directory!

### Checks by code reviewer(s):
* [x] Is the code clean of development artifacts e.g., unnecessary comments, prints, ...
* [x] The ticket goals for each associated issue are reached or problems are clearly addressed (i.e., a new issue was introduced).
* [x] There are appropriate **unit tests** and they pass.
* [x] The git history is clean and linearized for the merge request.
* [x] Coverage report for new code is acceptable.

mknaranja commented on cpp/tests/test_migration.cpp:
```diff
@@ -19,12 +19,12 @@
 */
```
pls rename to test_mobility

dabele replied:
makes sense, fixing.

mknaranja commented on cpp/models/secir/secir_parameters_io.h:
```diff
@@ -161,28 +168,28 @@ IOResult<Graph<Model, MigrationParameters>> read_graph(const std::string& direct
  */
 template <class Simulation>
-IOResult<void> write_single_run_result(const int run, const epi::Graph<epi::SimulationNode<Simulation>, epi::MigrationEdge>& graph)
+IOResult<void> write_single_run_result(const int run, const mio::Graph<mio::SimulationNode<Simulation>, mio::MigrationEdge>& graph)
```
maybe I was unable to find it, but this function is actually never tested, is it?

mknaranja commented on cpp/memilio/io/README.md:
```diff
@@ -0,0 +1,119 @@
+# MEmilio C++ IO
+
+This directory contains utility for reading and writing data from and to files in different formats. The main part is a serialization framework that can be used to define the structure of data without using a specific file format. There are implementations of the framework for different formats. The framework is described in detail below, also see the [serialization example](../../examples/serial ...
```
```suggestion
This directory contains utilities for reading and writing data from and to files in different formats. The main part is a serialization framework that can be used to define the structure of data without using a specific file format. There are implementations of the framework for different formats. The framework is described in detail below, also see the [serialization example](../../examples/serialize.cpp). 
```

joergbrech commented on cpp/memilio/CMakeLists.txt:
```diff
@@ -0,0 +1,88 @@
+    utils/date.cpp
+    utils/random_number_generator.h
+    utils/random_number_generator.cpp
+)
```
Just a comment (feel free to ignore). I find this list a bit intimidating. What do you think about either
1. using something like `file(GLOB_RECURSE source_list "*.cpp" "*.hpp")` , or
2. using `add_directory()` and amending the `source_list` by shorter lists in the `CMakeLists.txt` of the subdirectories io, utils, models, ...?

dabele replied:
> using something like file(GLOB_RECURSE source_list "*.cpp" "*.hpp")

this is not recommended (last I checked, maybe there is a way to make this work properly now?). The advantage of not having to change the CMakeLists is also the main drawback. Cmake doesn't detect new files automatically. So if you glob the sources like that, cmake configure must be run manually when a file is added or the new file is not built. This may be OK for the developer who adds the file, but it's weird for other developers. They will get unexpected linker errors if they pull changes from github because they don't know they need to run cmake. If the CMakeLists need to be modified every time a file is added, Cmake never needs to be run manually, it runs automatically when building. Since CI and local tests reliably detect when source files are missing from this list (the build just fails), the time saved for the person that adds files is not worth the additional work for everyone else.

> using add_directory() and amending the source_list by shorter lists in the CMakeLists.txt of the subdirectories io, utils, models, ..

I'm not opposed to that, although I wonder if this would be that much of an improvement? I think this means I'll always look in the wrong place, but maybe that's just me :smile:

joergbrech commented on cpp/models/secir/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+    implicit_euler.cpp
+    parameter_space.h
+    parameter_space.cpp
+    parameter_studies.h
```
I think something like `file(GLOB_RECURSE source_list "*.cpp" "*.h")` could make our life easier

dabele replied:
files are not needed anymore, only memilio/epidemiology/populations is used.

dabele commented on cpp/examples/graph.cpp:
```diff
@@ -17,9 +17,9 @@
-#include <epidemiology/migration/migration.h>
-#include <epidemiology/secir/seir.h>
-#include <epidemiology/model/simulation.h>
+#include "seir/seir.h"
```
no, as discussed. But it's just a question of cmake configuration, `models/seir/seir.h` would work as well if we changed the path in cmake. I don't think `models` adds anything useful.

closing all these discussions for now, but I'm open to using a different path.

dabele commented on cpp/models/abm/CMakeLists.txt:
```diff
@@ -0,0 +1,26 @@
+add_library(abm
```
I would like more order, sadly the order gets lost quickly in my experience.

there are a few headers missing in the list, fixing.

Btw. github can set a comment on a multi-line block, that would make answering of comments easier since all relevant lines are displayed in the overview.

dabele commented on cpp/thirdparty/README.md:
```diff
@@ -0,0 +1,20 @@
+# 3rd-Party Dependencies #
```
good point, there should be a short thing about the other bundled stuff, how it works, how to upgrade, even though it's easier than boost. But note that the dependencies themselves are explained (for users) in the main cpp README, this is just for developers.

dabele commented on .github/workflows/main.yml:
```diff
@@ -36,6 +36,17 @@ jobs:
           sanitize-addr: ${{ (matrix.compiler == 'gcc' && matrix.config == 'Debug' && matrix.version == 'latest') && 'ON' || 'OFF' }}
           sanitize-ub: ${{ (matrix.compiler == 'gcc' && matrix.config == 'Debug' && matrix.version == 'latest') && 'ON' || 'OFF' }}
 
+  build-linux-no-optional-deps:
```
for dependencies, the platform is more relevant than the compiler I think, that's why I did that. But maybe it's more consistent the other way.

dabele commented on cpp/memilio/io/io.h:
```diff
@@ -17,131 +17,22 @@
+#include "memilio/utils/compiler_diagnostics.h"
+#include "memilio/math/eigen_util.h"
+GCC_CLANG_DIAGNOSTIC(push)
+GCC_CLANG_DIAGNOSTIC(ignored "-Wc++17-extensions")
```
boost checks for specific features being available instead of checking the version of the standard. In this case, the `nodiscard` attribute was supported by gcc and clang as an extension long before c++17 came out, and it's really helpful to detect some errors. I think boost can be trusted to properly guard these extensions if they are not available, so I just disabled the warning here (but not globally). I don't think anything bad can happen if we forget this, worst case it becomes obsolete if we upgrade to c++17.

dabele commented on cpp/models/abm/migration_rules.cpp:
```diff
@@ -18,17 +18,17 @@
 * See the License for the specific language governing permissions and
```
local mobility is misleading, since in the context of the ABM, they are not local at all. We already use the terms local and global in the abm, local means for a single location, global is for the whole (ABM) world.

dabele commented on cpp/memilio/CMakeLists.txt:
```diff
@@ -0,0 +1,88 @@
+
+add_library(memilio
+    config.h
+    epidemiology/populations.h
```
obsolete populations files in secir have been removed.

dabele commented:
Rebased to resolve merge conflicts. All checks succesful.

Cleaned up commits to fix rename errors. empty cpp files still are sometimes not resolved correctly, but I don't think anything can be done about that. 

Commits are now all fairly self contained, I squashed a lot of the fix and review commits manually, so I think the PR can be merged without squashing, so more of the rename history is saved.
